### PR TITLE
capitalizeMsg function to correctly capitalize the first character an…

### PR DESCRIPTION
**Pull request by Horiodino**
handle error msf scenarios based

### What issues does this PR fix or reference?
https://github.com/devfile/devworkspace-operator/issues/1092

### PR Checklist

- [x] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

**Comment by Horiodino** [comment]
i was thinking in the dw structure we can add a context of type error if its for user we can set it as true and then at the end we can dw is for error handling and representing we can directly add function for that.
but for that we need to set context like if the error is for user then we need to set it as true and after that we can again set it to false ?

File: controllers/workspace/condition.go
Comment (AObuchow) [inline controllers/workspace/condition.go:136]: The go standard library actually has [cases](https://pkg.go.dev/golang.org/x/text/cases) for this use-case (handling unicode string capitalization).
Rather than re-invent the wheel, you could do something like:

```go
import (
	"golang.org/x/text/cases"
	"golang.org/x/text/language"
)
...

return cases.Title(language.English, cases.NoLower).String(msg)
```

**Review by AObuchow — COMMENTED** [review]
Thank you for your PR @Horiodino :)
It seems you're missing testing instructions. Could you please add some?
Have you tried building DevWorkspace Operator with your changes and [deploying it](https://github.com/devfile/devworkspace-operator/tree/main?tab=readme-ov-file#deploying-devworkspace-operator) to a testing cluster (like minikube for instance?)

I took a look through the codebase to try and find a user-facing error message that is not already being capitalized, and is simple enough to trigger. What I found was that finalize function (which gets called when a workspace is being terminated/deleted off the cluster) will try and [check](https://github.com/devfile/devworkspace-operator/blob/main/controllers/workspace/finalize.go#L102-L108) the workspace's storageProvisioner (i.e. storage type). The potential error that could happen at this step is [not being capitalized](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/storage/provisioner.go#L30).

You should be able to trigger this (normally not-capitalized error) by:
1. Creating a workspace with a valid storage strategy attribute set, e.g.:

```YAML
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: plain-devworkspace
spec:
  started: true
  routingClass: 'basic'
  template:
    attributes:
      controller.devfile.io/storage-type: per-user
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
```
2. Let the workspace startup successfully
3. Modify the workspace's storage-type attribute to an invalid value
4. Delete the workspace object off the cluster, i.e. `kubectl delete dw plain-devworkspace -n <your-workspace's-namespace>`
5. Check that the workspace status error message is capitalized by checking the devworkspace object or doing a `kubectl get dw -n <your-workspace's-namespace>`

**Comment by AObuchow** [comment]
> i was thinking in the dw structure we can add a context of type error if its for user we can set it as true and then at the end we can dw is for error handling and representing we can directly add function for that. but for that we need to set context like if the error is for user then we need to set it as true and after that we can again set it to false ?

Anytime we're setting the workspace status, such as [here](https://github.com/devfile/devworkspace-operator/blob/main/controllers/workspace/devworkspace_controller.go#L585), it is already implied that it is a user-facing message. As the workspace status will be [provided in the workspace object on the cluster](https://github.com/devfile/devworkspace-operator/blob/main/controllers/workspace/status.go#L66). So there should be no need for this additional context IMO.

**Comment by Horiodino** [comment]
sorry for making so long , the `make docker` was taking so much time (might be isp problem) .
on running 
```
kubectl patch dw plain-devworkspace --type=merge -p '{"spec":{"template":{"attributes":{"controller.devfile.io/storage-type": "invalid-type"}}}}'
```
```
kubectl get dw -w
NAME                 DEVWORKSPACE ID             PHASE     INFO
plain-devworkspace   workspace147061c666194f78   Running   Workspace is running
plain-devworkspace   workspace147061c666194f78   Running   Workspace is running
plain-devworkspace   workspace147061c666194f78   Failing   Error Provisioning Storage: Configured Storage Type Not Supported
plain-devworkspace   workspace147061c666194f78   Failing   Error Provisioning Storage: Configured Storage Type Not Supported
plain-devworkspace   workspace147061c666194f78   Failing   Error Provisioning Storage: Configured Storage Type Not Supported
plain-devworkspace   workspace147061c666194f78   Failing   Error Provisioning Storage: Configured Storage Type Not Supported
plain-devworkspace   workspace147061c666194f78   Failed    Error Provisioning Storage: Configured Storage Type Not Supported```

**Comment by AObuchow** [comment]
@Horiodino thank you for the follow-up. Could you please run a `go mod tidy` on your fork and commit the changes? This should fix the failing GH Action.

You might also need to do a `go install golang.org/x/tools/cmd/goimports@latest` followed by a `make fmt`

**Comment by Horiodino** [comment]
@AObuchow

**Comment by AObuchow** [comment]
@Horiodino sorry for the delay, I'm planning to continue reviewing your PR early this week.

**Review by AObuchow — CHANGES_REQUESTED** [review]
Sorry for the delay in following up on this @Horiodino.

I made a silly mistake in my [original review](https://github.com/devfile/devworkspace-operator/pull/1276/commits/e75892616cdb50fc3679543052cb5c71ca17f472#r1651703541): `cases.Title(language.English, cases.NoLower)` actually sets every word in a string to start with an upper case character, which is not what we want here.

You can remove commits https://github.com/devfile/devworkspace-operator/pull/1276/commits/f56a79cfd16957ca7204402d47ab622c84980d49 & https://github.com/devfile/devworkspace-operator/pull/1276/commits/c431b0ea38ee5ef12ed554ad5016adccfd12766f from your PR and just keep https://github.com/devfile/devworkspace-operator/pull/1276/commits/e75892616cdb50fc3679543052cb5c71ca17f472. I'll add some extra comments on that commit. 

Also for [testing](https://github.com/devfile/devworkspace-operator/pull/1276#issuecomment-2210852153), you should be running 
```
kubectl patch dw plain-devworkspace -n $NAMESPACE --type=merge -p '{"spec":{"template":{"attributes":{"controller.devfile.io/storage-type": "invalid-type"}}}}' && kubectl delete dw plain-devworkspace -n $NAMESPACE
```
as the workspace needs to be deleted immediately after setting an invalid storage type.

The output of `kubectl get dw -n $NAMESPACE -w` should include something like:
```
plain-devworkspace   workspace3a2024c23f13435f   Error         Configured Storage Type Not Supported
```

File: controllers/workspace/condition.go
Comment (AObuchow) [inline controllers/workspace/condition.go]: We usually check for empty strings directly, rather than checking the length of the string. So this can be rewritten as:
```go
if msg == "" {
  return msg
}
```

File: controllers/workspace/condition.go
Comment (AObuchow) [inline controllers/workspace/condition.go]: I think this function name could be unabbreviated as `capitalizeMessage(msg string)`.

Also, it's worth adding a simple function doc, something like:
`// Capitalizes the first character in the given string.`

**Comment by Horiodino** [comment]
working as expected 
```plain-devworkspace   workspaced049889b5c554e91   Error         Configured storage type not supported```

**Review by AObuchow — APPROVED** [review]
LGTM. @dkwon17 any thoughts? I'll wait for you to give this a look before we merge.

**Review by dkwon17 — APPROVED** [review]
Sorry, just getting to this now,

LGTM, let's have it for DWO 0.31.x

**Comment by AObuchow** [comment]
@Horiodino This looks just about good to merge finally. @Horiodino could you please squashing both of your commits and perhaps rewording the commit message to "feat: ensure all DevWorkspace condition messages are capitalized" (or something similar)?

Thank you :)

**Comment by AObuchow** [comment]
@Horiodino Sorry for being so picky, but your commit now has the author sign-off twice in the description, and it's also lacking a reference to the bug it's resolving. Could you please modify your commit message to:

```
feat: ensure all DevWorkspace condition messages are capitalized

Fix #1092 

Signed-off-by: Horiodino <holiodin@gmail.com>
```

**Comment by Horiodino** [comment]
> @Horiodino Sorry for being so picky, but your commit now has the author sign-off twice in the description, and it's also lacking a reference to the bug it's resolving. Could you please modify your commit message to:
> 
> ```
> feat: ensure all DevWorkspace condition messages are capitalized
> 
> Fix #1092 
> 
> Signed-off-by: Horiodino <holiodin@gmail.com>
> ```

np , done :+1:

**Comment by AObuchow** [comment]
@Horiodino Thank you! Merging, great work :)

**Pull request merged**
